### PR TITLE
fix If ToS file is not set, customer needs to accept Terms on each checkout

### LINKED ISSUE #10675: If ToS file is not set, customer needs to accept Terms on each checkout
## Description

If ToS file is not set, customer needs to accept Terms on each checkout

## Expected Behavior

Even when a ToS file is not set the customer should need to accept Terms on the first checkout only

## Actual Behaviour

If ToS file is not set, customer needs to accept Terms on each checkout

## Steps to Reproduce

1. As a superadmin, visit `/admin/terms_of_service_files` and make sure only the default ToS file is available
2. As a customer, checkout twice: notice that on both checkouts it is necessary to check the checkbox

## Animated Gif/Screenshot

On the second checkout:

if a ToS file is set (staging-UK, left) - correct behavior
if no ToS file is set (staging-FR, right) - terms need to be accepted every time

image

## Workaround

Set a ToS file, as superadmin.

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: 4.3.2 (introduced on a previous version)
* Browser name and version: Firefox 110
* Operating System and version (desktop or mobile): Ubuntu 22.04

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #10675

if ToS file is not set, `TermsOfServiceFile.updated_at`  returns `Time.now`.
This makes `TermsOfService.tos_accepted?` always return `false` on the described context  (ToS file not set + distributor doesn't require ToS)

#### What should we test?

1. As a superadmin, visit /admin/terms_of_service_files and make sure only the default ToS file is available
2. As a customer, checkout twice: notice that on both checkouts it is necessary to check the checkbox

#### Release notes

Changelog Category: User facing changes | Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by abdellani** [comment]
Not sure why it's failing, since no error was raised.
All the tests on that file passed locally.
image

**Comment by rioug** [comment]
Your forgot to remove a `pending` in `spec/system/consumer/split_checkout_spec.rb` and apparently when a pending test becomes successful it fails the test run hence why it's failing.

**Review by mkllnk — APPROVED** [review]

**Comment by mkllnk** [comment]
I updated the spec.

**Review by rioug — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @abdellani,
Thanks for working on this one! :pray: 

With no ToS file set:
image

I checked out with a new user:
- I can confirm that the new customer has to accept the ToS only during the first checkout. After that the box is already checked. :heavy_check_mark: 

Trying this with an existing user it was not possible to test this because the checkbox was checked on the first checkout after deleting the ToS already. I don't think is a good behaviour. :x:
If the ToS file gets deleted, then the default ToS template is used, so the ToS changed for the customer. It should be required to actively accept those default ToS by leaving the box unchecked on the first checkout after deleting the ToS file.

I hope you understand what I mean. Let me know if it's unclear.

I will move this one back to In Dev to fix the last bit.

Thanks!

**Comment by drummer83** [comment]
Yes, almost. D1 can't set the ToS, the instance manager has to do this.
D1 can only set Term And Conditions - which is similar but not what the issue was about.

**Comment by abdellani** [comment]
> Yes, almost. D1 can't set the ToS, the instance manager has to do this. D1 can only set Term And Conditions - which is similar but not what the issue was about.

Yes,  and if the customer accepts TaC, he'll not need to accept ToS.


Now, here's what I understood from the code.

We are only storing the time, let's call that time `T1`,  when the customer accepts the ToS/ToC. We are not storing which ToS/ToC was accepted.

1. If the distributor sets a ToC, we compare `T1` vs the creation time (`CT_TOC`) of the ToC => if `T1` comes after creation time, this means the customer accepted the `ToC`.
2. if the distributor didn't set ToC, we check if a default ToS exists

    * if the ToS exists, we do the same thing between the creation time of ToS `CT_TOS`  vs `T1`
    * if the ToS doesn't exist, we consider that the customer already accepted the ToS :point_right: This is what I solved on this PR. 


The bug you spotted after the tests will appear when:  
T1 > CT_TOC && T1 > CT_TOS
We don't need to care about the order between CT_TOC and CT_TOS.

Under those conditions, if the TOC is removed, we'll still have
T1 > CT_TOS => This will mean that the user already accepted the ToS.

note:
X > Y => X is more recent than Y


To me the problem is clear. I can see different approaches to solve it but I'm not sure which one is more appropriate in our context.

* We can add a flag to say if the document that was accepted was a ToC or ToS.
* We can set `T1` to `nil` if the ToS/ToC is added/removed.

**Comment by drummer83** [comment]
> Yes, and if the customer accepts TaC, he'll not need to accept ToS.

I can't follow this. If the instance has configured the settings so that the customers have to accept the ToS and the enterprise has uploaded TaC, then the customer needs to accept both at the same time:
image

> - We can add a flag to say if the document that was accepted was a ToC or ToS.
> 
> - We can set `T1` to `nil` if the ToS/ToC is added/removed.

From my non-technical point of view I would think that setting T1 to nil would be an easy and good solution. If you are not sure maybe let's get feedback from @openfoodfoundation/train-drivers-product-owners and/or @openfoodfoundation/core-devs?

**Comment by RachL** [comment]
Just to be sure: what are we calling default ToS? To me, if the shop does not have a ToS file, there is no other file that can be shown by default...

**Comment by drummer83** [comment]
The enterprise has a TaC file. If this gets deleted, then nothing will be shown, the customer doesn't have to agree to it.
image

The instance has a ToS file. If this is deleted, then a default file will be shown.
A file uploaded:
image

The file deleted:
image

The default file is shown if no file was uploaded and still the agreement is required by the instance:
image

**Comment by abdellani** [comment]
> > Yes, and if the customer accepts TaC, he'll not need to accept ToS.
> 
> I can't follow this. If the instance has configured the settings so that the customers have to accept the ToS and the enterprise has uploaded TaC, then the customer needs to accept both at the same time: image
> 

No, he needs to accept only one document.
if TaC is set, it'll be the first priority
if it's not, then we check ToS.

> > * We can add a flag to say if the document that was accepted was a ToC or ToS.
> > * We can set `T1` to `nil` if the ToS/ToC is added/removed.
> 
> From my non-technical point of view I would think that setting T1 to nil would be an easy and good solution. If you are not sure maybe let's get feedback from @openfoodfoundation/train-drivers-product-owners and/or @openfoodfoundation/core-devs?

Ok, I'll wait for the feedback.

**Comment by mkllnk** [comment]
The default Terms of Service are an old file from the Australian instance. It was kept in place for backwards compatibility to start with but it's not valid for any other instance and Australia updated their ToS in the meantime. We should probably remove it from the code.

If I understand @drummer83 correctly, you were concerned that the default ToS was seen as accepted when the previous file was deleted. I think that this is an invalid scenario. A platform without its own ToS file doesn't have a valid ToS file and the config of requiring ToS without providing a file is invalid.

Instead of adding more code, flags and checks, we should just remove the default and not show any ToS requirement in the frontend unless a file has been uploaded.

**Comment by RachL** [comment]
Aligned with Maikel's proposal's :+1:

File: app/models/spree/app_configuration.rb [hunk 1]
```diff
@@ -105,7 +105,6 @@ class AppConfiguration < Preferences::Configuration
     preference :embedded_shopfronts_whitelist, :text, default: nil
 
     # Legal Preferences
-    preference :footer_tos_url, :string, default: "/Terms-of-service.pdf"
```
Comment (mkllnk) [inline app/models/spree/app_configuration.rb:108]: Oh, this is going a step further than I anticipated. I was talking about the default being obsolete. But some instances may still use this old setting with valid URLs. I forgot that option. The spree_preferences table does have an updated_at attribute which we could use. But, hey, let's check first if we can get rid of the altogether as you are suggesting now.

```
$ sql='SELECT * FROM spree_preferences WHERE key = $$/spree/app_configuration/footer_tos_url$$ LIMIT 1;'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
  id   |         value         |        created_at         |         updated_at         |                   key                   | value_type 
-------+-----------------------+---------------------------+----------------------------+-----------------------------------------+------------
 13790 | /Terms-of-service.pdf | 2019-01-25 04:55:28.99796 | 2020-02-24 22:31:24.286698 | /spree/app_configuration/footer_tos_url | string
(1 row)
openfoodnetwork.org.nz | CHANGED | rc=0 >>
 id |                    value                     |         created_at         |         updated_at         |                   key                   | value_type 
----+----------------------------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 31 | https://openfoodnetwork.nz/terms-of-service/ | 2019-09-24 07:41:12.235821 | 2019-12-06 09:57:28.479701 | /spree/app_configuration/footer_tos_url | string
(1 row)
openfoodnetwork.ie | CHANGED | rc=0 >>
 id |                   value                    |         created_at         |         updated_at         |                   key                   | value_type 
----+--------------------------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 30 | --- https://about.openfoodnetwork.ie/terms+| 2020-03-28 18:40:15.753541 | 2020-08-21 20:56:33.074604 | /spree/app_configuration/footer_tos_url | string
    | ...                                       +|                            |                            |                                         | 
    |                                            |                            |                            |                                         | 
(1 row)
openfoodnetwork.org.uk | CHANGED | rc=0 >>
  id  |                                       value                                        |         created_at         |         updated_at         |                   key                   | value_type 
------+------------------------------------------------------------------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 8538 | https://drive.google.com/file/d/1ne3GqiXnzTw6tWD484agzWqhlerJzH8p/view?usp=sharing | 2018-09-08 10:48:37.934328 | 2020-06-04 07:57:29.594535 | /spree/app_configuration/footer_tos_url | string
(1 row)
openfoodnetwork.ca | CHANGED | rc=0 >>
  id  |            value            |         created_at         |         updated_at         |                   key                   | value_type 
------+-----------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 2023 | --- "/Terms-of-service.pdf"+| 2018-10-08 03:20:50.636275 | 2020-11-03 21:29:56.958475 | /spree/app_configuration/footer_tos_url | string
      |                             |                            |                            |                                         | 
(1 row)
app.katuma.org | CHANGED | rc=0 >>
 id  |            value            |         created_at         |         updated_at         |                   key                   | value_type 
-----+-----------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 926 | --- "/Terms-of-service.pdf"+| 2018-12-03 13:16:17.696466 | 2020-11-06 18:08:07.512145 | /spree/app_configuration/footer_tos_url | string
     |                             |                            |                            |                                         | 
(1 row)
csicsoka.org | CHANGED | rc=0 >>
 id | value | created_at | updated_at | key | value_type 
----+-------+------------+------------+-----+------------
(0 rows)
openfoodnetwork.de | CHANGED | rc=0 >>
 id  |            value            |         created_at         |         updated_at         |                   key                   | value_type 
-----+-----------------------------+----------------------------+----------------------------+-----------------------------------------+------------
 300 | --- "/Terms-of-service.pdf"+| 2019-04-16 10:19:36.117812 | 2020-12-04 18:43:16.129764 | /spree/app_configuration/footer_tos_url | string
     |                             |                            |                            |                                         | 
(1 row)
openfoodnetwork.net | CHANGED | rc=0 >>
 id  |            value            |         created_at         |        updated_at         |                   key                   | value_type 
-----+-----------------------------+----------------------------+---------------------------+-----------------------------------------+------------
 527 | --- "/Terms-of-service.pdf"+| 2018-11-16 03:58:49.307589 | 2020-11-16 13:25:56.34786 | /spree/app_configuration/footer_tos_url | string
     |                             |                            |                           |                                         | 
(1 row)
openfoodnetwork.be | CHANGED | rc=0 >>
 id |         value         |         created_at         |         updated_at         |                   key                   | value_type 
----+-----------------------+----------------------------+----------------------------+-----------------------------------------+------------
 41 | /Terms-of-service.pdf | 2018-10-29 20:04:44.172964 | 2019-10-28 13:59:52.575425 | /spree/app_configuration/footer_tos_url | string
(1 row)
coopcircuits.fr | CHANGED | rc=0 >>
  id  |                           value                           |         created_at         |        updated_at         |                   key                   | value_type 
------+-----------------------------------------------------------+----------------------------+---------------------------+-----------------------------------------+------------
 3992 | --- https://apropos.coopcircuits.fr/informations-legales/+| 2018-10-09 13:03:37.022249 | 2020-11-16 09:32:46.59564 | /spree/app_configuration/footer_tos_url | string
      | ...                                                      +|                            |                           |                                         | 
      |                                                           |                            |                           |                                         | 
(1 row)
```


```
$ sql='SELECT COUNT(*) FROM terms_of_service_files'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
 count 
-------
     2
(1 row)
openfoodnetwork.org.nz | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.ca | CHANGED | rc=0 >>
 count 
-------
     1
(1 row)
openfoodnetwork.org.uk | CHANGED | rc=0 >>
 count 
-------
     1
(1 row)
app.katuma.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
csicsoka.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.de | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.net | CHANGED | rc=0 >>
 count 
-------
     1
(1 row)
coopcircuits.fr | CHANGED | rc=0 >>
 count 
-------
     1
(1 row)
openfoodnetwork.be | CHANGED | rc=0 >>
 count 
-------
     1
(1 row)
openfoodnetwork.ie | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
```

Instances without uploaded file:
- openfoodnetwork.org.nz -> https://openfoodnetwork.nz/terms-of-service/
- app.katuma.org -> /Terms-of-service.pdf
- csicsoka.org -> nil, default
- openfoodnetwork.de -> /Terms-of-service.pdf
- openfoodnetwork.ie -> https://about.openfoodnetwork.ie/terms

So we have two instances which would need to upload their terms of service as PDF.

Maybe a question for @openfoodfoundation/train-drivers-product-owners: Do we want to continue to support a URL as platform terms of service (2 instances) or do we want all instances to use an uploaded PDF file (6 instances)?

Only PDF is less code to maintain but possibly not as good UX.

**Comment by RachL** [comment]
Update: IE uploaded their file, waiting for confirmation from NZ 👍

**Review by mkllnk — APPROVED** [review]
Cool, let's go ahead with this.

**Review by rioug — APPROVED** [review]
Looks good :+1:

**Comment by drummer83** [comment]
Hi @abdellani,
Thanks for fixing this.

It was interesting for me to see that we are keeping a history of ToS files. Deleting one file activates the previously uploaded file until all files have been deleted. I think this is ok and not a problem really. :+1: 

Here are some scenarios I have tested. Columns 1, 2 and 4 are the settings in the instance configuration, columns 3 and 5 are the resulting app behaviour for users.

| ToS file uploaded? | Enterprises must accept ToS (during registration) | Result | Shoppers must accept ToS (during checkout) | Result |
| - | - | - | - | - |
| No (during staging) | Yes | :x: A link without file is shown (see video 1). | Yes | :heavy_check_mark: No acceptance required. |
| No | No | :heavy_check_mark: No acceptance required. | No | :heavy_check_mark: No acceptance required. |
| Yes (new file) | Yes | :heavy_check_mark: Acceptance required. | Yes | :heavy_check_mark: Acceptance required. | 
| Yes | No | :heavy_check_mark: No acceptance required. | No | :heavy_check_mark: No acceptance required. | 
| Yes (still same file) | Yes | :heavy_check_mark: Acceptance required (for every enterprise). | Yes | :heavy_check_mark: Acceptance already activated. |
| Yes (file updated) | Yes | :heavy_check_mark: Acceptance required. | Yes | :heavy_check_mark: New acceptance required. |
| No (all files deleted) | Yes | :x: A link without file is shown (see video 1). | Yes | :heavy_check_mark: No acceptance required. | 

Video 1:

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/a16b1739-cb49-4a01-84ba-7ca92f63b00b

### Conclusion
As you can see we have unfortunate behaviour when no file is uploaded but the instance has the switch 'Enterprises must accept ToS' activated. I think this is a bug, but probably s4 only, so I will create an issue but we don't have to fix it right away. We could however inform the instances that they should switch it off if they have no file uploaded - which they probably already have.

In summary I think this is ready to go! :partying_face:

**Comment by RachL** [comment]
NZ confirmed as well :+1:
